File: client/archive/ios/BriskVTU/Features/Coupons/CouponAPI.swift
import Foundation

struct CouponResponse: Codable, Identifiable {
    let id: String
    let code: String
    let name: String
    let amountPerUse: Int
    let maxUses: Int
    let remainingUses: Int
    let status: String
    let currency: String
    let expiresAt: Date?
    let createdAt: Date
    let creatorUserId: String
}

struct CouponRedemptionResponse: Codable, Identifiable {
    let id: String
    let couponId: String
    let redeemerUserId: String
    let amount: Int
    let status: String
    let createdAt: Date
}

struct CreateCouponRequest: Codable {
    let amount_per_use: Int
    let max_uses: Int
    let currency: String
    let name: String?
    let expires_at: String?
    let idempotency_key: String
}

struct RedeemCouponRequest: Codable {
    let code: String
    let idempotency_key: String
}

struct UpdateCouponRequest: Codable {
    let name: String?
    let expires_at: String?
}

final class CouponAPI {
    static let shared = CouponAPI()
    private let client = APIClient.shared
    
    func createCoupon(request: CreateCouponRequest) async throws -> CouponResponse {
        let data = try JSONEncoder().encode(request)
        let res: BaseResponse<CouponResponse> = try await client.request(endpoint: "/coupons", method: "POST", body: data)
        guard let coupon = res.data else { throw APIError.noData }
        return coupon
    }
    
    func redeemCoupon(request: RedeemCouponRequest) async throws -> CouponRedemptionResponse {
        let data = try JSONEncoder().encode(request)
        let res: BaseResponse<CouponRedemptionResponse> = try await client.request(endpoint: "/coupons/redeem", method: "POST", body: data)
        guard let redemption = res.data else { throw APIError.noData }
        return redemption
    }
    
    func listMyCoupons() async throws -> [CouponResponse] {
        let res: BaseResponse<[CouponResponse]> = try await client.request(endpoint: "/coupons", method: "GET")
        return res.data ?? []
    }
    
    func pauseCoupon(id: String) async throws {
        let _: BaseResponse<EmptyResponse> = try await client.request(endpoint: "/coupons/\(id)/pause", method: "POST")
    }
    
    func resumeCoupon(id: String) async throws {
        let _: BaseResponse<EmptyResponse> = try await client.request(endpoint: "/coupons/\(id)/resume", method: "POST")
    }
    
    func revokeCoupon(id: String) async throws {
        let _: BaseResponse<EmptyResponse> = try await client.request(endpoint: "/coupons/\(id)/revoke", method: "POST")
    }
}

File: client/archive/android/app/src/main/java/com/solutioncrafts/BriskVTU/feature/coupons/CouponApi.kt
package com.solutioncrafts.BriskVTU.feature.coupons

import kotlinx.serialization.Serializable
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path

@Serializable
data class CouponResponse(
    val id: String,
    val code: String,
    val name: String,
    val amountPerUse: Int,
    val maxUses: Int,
    val remainingUses: Int,
    val status: String,
    val currency: String,
    val expiresAt: String? = null,
    val createdAt: String,
    val creatorUserId: String
)

@Serializable
data class CouponRedemptionResponse(
    val id: String,
    val couponId: String,
    val redeemerUserId: String,
    val amount: Int,
    val status: String,
    val createdAt: String
)

@Serializable
data class CreateCouponRequest(
    val amount_per_use: Int,
    val max_uses: Int,
    val currency: String = "NGN",
    val name: String? = null,
    val expires_at: String? = null,
    val idempotency_key: String
)

@Serializable
data class RedeemCouponRequest(
    val code: String,
    val idempotency_key: String
)

@Serializable
data class UpdateCouponRequest(
    val name: String? = null,
    val expires_at: String? = null
)

@Serializable
data class BaseResponse<T>(
    val status: String,
    val data: T? = null,
    val message: String? = null
)

interface CouponApi {
    @POST("coupons")
    suspend fun createCoupon(@Body request: CreateCouponRequest): BaseResponse<CouponResponse>

    @POST("coupons/redeem")
    suspend fun redeemCoupon(@Body request: RedeemCouponRequest): BaseResponse<CouponRedemptionResponse>

    @GET("coupons")
    suspend fun listMyCoupons(): BaseResponse<List<CouponResponse>>

    @GET("coupons/{id}")
    suspend fun getCouponDetails(@Path("id") id: String): BaseResponse<CouponResponse>

    @PATCH("coupons/{id}")
    suspend fun updateCoupon(
        @Path("id") id: String,
        @Body request: UpdateCouponRequest
    ): BaseResponse<CouponResponse>

    @POST("coupons/{id}/pause")
    suspend fun pauseCoupon(@Path("id") id: String): BaseResponse<Unit>

    @POST("coupons/{id}/resume")
    suspend fun resumeCoupon(@Path("id") id: String): BaseResponse<Unit>

    @POST("coupons/{id}/revoke")
    suspend fun revokeCoupon(@Path("id") id: String): BaseResponse<Unit>
}
